# ☕ Laporan Pagi Raka: Materi Lengkap
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/thosangs/python_lecture/blob/main/notebooks/01_materi_kopi_senja.ipynb)

Notebook materi **Pertemuan 1 · Python Fundamentals & Logic**. Semua cell **sudah dijalankan**, jadi kamu bisa membaca hasilnya langsung, termasuk error yang **sengaja** dibuat supaya kita belajar membacanya.

> **Ceritanya:** Raka adalah data analyst di **Kopi Senja** (3 cabang: Jakarta, Bandung, Surabaya). Tiap pagi Raka butuh **65 menit** untuk menyusun laporan penjualan secara manual, dan pernah salah ketik satu nol. Hari ini kita bantu Raka membuat **laporan otomatis dalam 1 detik**.

| Section | Isi |
|---|---|
| **00 · Demo Final** | Hasil akhir yang akan kita bangun |
| **Bab 3 · Halo Kopi Senja** | `print` pertama, Python itu *interpreted*, error pertama |
| **Bab 4 · Toples Berlabel** | Komentar, `print`, variabel, aturan nama, `del` |
| **Bab 5 · Ngitung Omzet** | `int`, `float`, operator, fungsi angka, boolean |
| **Bab 6 · Beresin Nama Menu** | String, indexing, slicing, method, casting |
| **Bab 7 · Wadah Banyak Barang** | `list`, `tuple`, `set`, `dict` |
| **Bab 8 · Laporan buat Bu Sari** | Escape character, f-string, `input()` |
| **Final · Laporan Harian v1.0** | Gabungkan semuanya |

📊 Slide: https://thosangs.github.io/python_lecture/ · ✍️ Latihan: `02_latihan_peserta.ipynb` · 🔑 Kunci: `03_kunci_jawaban.ipynb`

> 💡 Mau coba sendiri? **File → Save a copy in Drive**, lalu **Runtime → Run all**.

## 00 · Demo Final
Ini laporan yang akan **kamu bikin sendiri** di akhir kelas. Jalankan cell di bawah, lalu isi nama dan tanggal saat diminta.

In [1]:
# Laporan Harian Kopi Senja v1.0

# --- 1. Konstanta: nilai yang nggak berubah ---           (Bab 4)
NAMA_TOKO = "Kopi Senja"
TARGET_HARIAN = 10_000_000
JAM_OPERASIONAL = (7, 22)                                 # (Bab 7: tuple)

# --- 2. Input dari orang yang menjalankan script ---      (Bab 8 + Bab 6)
nama_analis = input("Nama analis: ").strip().title()
tanggal = input("Tanggal laporan (DD-MM-YYYY): ").strip()

# --- 3. Data hari ini (ceritanya hasil download kasir) --- (Bab 7: list)
cabang = ["Jakarta", "Bandung", "Surabaya"]
omzet = [4_250_000, 2_875_000, 3_520_000]
transaksi = [170, 125, 160]
menu_terlaris_mentah = "   es KOPI susu gula AREN  "
kode_trx_terakhir = "TRX-20261005-JKT-0170"

# --- 4. Olah data ---                                      (Bab 5)
total_omzet = sum(omzet)
total_transaksi = sum(transaksi)
rata_rata = round(total_omzet / total_transaksi)
selisih = max(omzet) - min(omzet)
persen_target = total_omzet / TARGET_HARIAN * 100
target_tercapai = total_omzet >= TARGET_HARIAN

menu_terlaris = menu_terlaris_mentah.strip().title()      # (Bab 6)
kode_cabang = kode_trx_terakhir[13:16]
nomor_trx = int(kode_trx_terakhir[-4:])

# --- 5. Cetak laporan ---                                  (Bab 8)
garis = "=" * 44
garis_tipis = "-" * 44

print(garis)
print(f"LAPORAN HARIAN {NAMA_TOKO.upper()}".center(44))
print(garis)
print(f"Tanggal          : {tanggal}")
print(f"Analis           : {nama_analis}")
print(f"Jam operasional  : {JAM_OPERASIONAL[0]}.00 - {JAM_OPERASIONAL[1]}.00")
print(garis_tipis)
print("Omzet per cabang")
print(f"  {cabang[0]}\t: Rp{omzet[0]:,} ({transaksi[0]} trx)".replace(",", "."))
print(f"  {cabang[1]}\t: Rp{omzet[1]:,} ({transaksi[1]} trx)".replace(",", "."))
print(f"  {cabang[2]}\t: Rp{omzet[2]:,} ({transaksi[2]} trx)".replace(",", "."))
print(garis_tipis)
print(f"Total omzet      : Rp{total_omzet:,}".replace(",", "."))
print(f"Total transaksi  : {total_transaksi}")
print(f"Rata-rata/trx    : Rp{rata_rata:,}".replace(",", "."))
print(f"Selisih max-min  : Rp{selisih:,}".replace(",", "."))
print(f"Menu terlaris    : {menu_terlaris}")
print(f"Trx terakhir     : #{nomor_trx} di cabang {kode_cabang}")
print(garis_tipis)
print(f"Target harian    : Rp{TARGET_HARIAN:,}".replace(",", "."))
print(f"Capaian          : {persen_target:.2f}%")
print(f"Target tercapai? : {target_tercapai}")
print(garis)
print("Laporan ini dibuat otomatis oleh Python ☕")

Nama analis:   raka pratama 
Tanggal laporan (DD-MM-YYYY): 05-10-2026
         LAPORAN HARIAN KOPI SENJA          
Tanggal          : 05-10-2026
Analis           : Raka Pratama
Jam operasional  : 7.00 - 22.00
--------------------------------------------
Omzet per cabang
  Jakarta	: Rp4.250.000 (170 trx)
  Bandung	: Rp2.875.000 (125 trx)
  Surabaya	: Rp3.520.000 (160 trx)
--------------------------------------------
Total omzet      : Rp10.645.000
Total transaksi  : 455
Rata-rata/trx    : Rp23.396
Selisih max-min  : Rp1.375.000
Menu terlaris    : Es Kopi Susu Gula Aren
Trx terakhir     : #170 di cabang JKT
--------------------------------------------
Target harian    : Rp10.000.000
Capaian          : 106.45%
Target tercapai? : True
Laporan ini dibuat otomatis oleh Python ☕


Kode ini cuma ±50 baris, dan **setiap barisnya** memakai konsep yang kita pelajari hari ini. Yuk mulai dari nol. 👇

## Bab 3 · Halo Kopi Senja
### Program pertama
Klik cell, lalu tekan **Shift + Enter** untuk menjalankan.

In [2]:
print("Halo, Kopi Senja!")

Halo, Kopi Senja!


### Python itu *interpreted*: dibaca baris per baris
Kayak barista yang meracik sambil membaca resep. **Tebak dulu:** baris 3 ada typo. Berapa baris yang tercetak?

In [3]:
print("1. Download data Jakarta... beres")
print("2. Download data Bandung... beres")
prnt("3. Download data Surabaya...")
print("4. Kirim laporan ke Bu Sari")

1. Download data Jakarta... beres
2. Download data Bandung... beres


NameError: name 'prnt' is not defined

Baris 1–2 jalan, baris 3 error, baris 4 **tidak pernah dijalankan**.

**Cara baca error dalam 3 langkah**
1. 👇 **Baca baris paling bawah dulu**: jenis error + pesannya
2. 🔍 **Cari nomor baris / tanda panah**: di baris mana error terjadi
3. 🤔 **Bandingkan dengan maksudmu**: typo? lupa tanda kutip? kurung belum ditutup?

> **NameError** = "Aku nggak kenal nama ini." Biasanya karena typo, lupa tanda kutip, atau cell yang membuat variabelnya belum dijalankan.

### Bonus: Zen of Python
Prinsip desain Python. Perhatikan baris *"Readability counts."*

In [4]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Bab 4 · Toples Berlabel
### Komentar: catatan untuk manusia
Semua yang diawali `#` dilewati Python. Shortcut: **Cmd + /** (Mac) atau **Ctrl + /** (Windows).

In [5]:
# Laporan harian Kopi Senja
# Dibuat oleh: Raka
print("Laporan dimulai")   # komentar juga bisa di ujung baris
# print("baris ini nggak dijalankan")

Laporan dimulai


### `print()` dan konsep function
**Function** = mesin kopi: bahan masuk lewat kurung `( )`, hasil keluar. **Built-in function** = mesin bawaan Python.

In [6]:
print("Laporan Kopi Senja")
print(3)
print("Jumlah cabang:", 3)    # beberapa nilai, dipisah koma
print()                       # baris kosong
print("selesai")

Laporan Kopi Senja
3
Jumlah cabang: 3

selesai


Teks **wajib** pakai tanda kutip. Tanpa kutip, Python mengira `Kopi` adalah nama variabel:

In [7]:
print(Kopi)

NameError: name 'Kopi' is not defined

### Variabel = toples berlabel
`nama_variabel = nilai`. Tanda `=` artinya **MASUKKAN KE**, bukan *sama dengan*. Baca dari kanan ke kiri.

In [8]:
nama_toko = "Kopi Senja"
jumlah_cabang = 3
omzet_jakarta = 4250000

print(nama_toko)
print(omzet_jakarta)

Kopi Senja
4250000


**Tebak output:** (A) 10 · (B) 12 · (C) Error

In [9]:
stok_cup = 10
stok_cup = stok_cup - 3
stok_cup = stok_cup + 5
print(stok_cup)

12


Di matematika `x = x + 1` itu mustahil. Di Python: *ambil isi toples, tambah 1, masukkan lagi*.

In [10]:
x = 5
x = x + 1
print(x)

6


### Isi toples bisa diganti (*dynamic typing*)
Python menebak tipe data dari isinya. Boleh ganti tipe, tapi **bikin bingung**, sebaiknya dihindari.

In [11]:
omzet_jakarta = 4250000
print(type(omzet_jakarta))

omzet_jakarta = "empat juta"
print(type(omzet_jakarta))

<class 'int'>
<class 'str'>


### Aturan wajib kasih nama
Kalau dilanggar → `SyntaxError` dan **seluruh cell** tidak dijalankan. Tiap contoh di bawah ada di cell sendiri.

**Tidak boleh diawali angka:**

In [12]:
2cabang = 3

SyntaxError: invalid decimal literal (107630894.py, line 1)

**Tidak boleh pakai spasi:**

In [13]:
omzet bandung = 875000

SyntaxError: invalid syntax (1134104170.py, line 1)

**Hanya huruf, angka, dan `_`** (tanda `-` dibaca sebagai pengurangan):

In [14]:
omzet-surabaya = 3520000

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (2782693036.py, line 1)

**Bukan keyword Python:**

In [15]:
class = "A"

SyntaxError: invalid syntax (1184760323.py, line 1)

**Case sensitive**: `Omzet` dan `omzet` itu beda:

In [16]:
Omzet = 100
print(omzet)

[4250000, 2875000, 3520000]


Daftar keyword (kata yang sudah dipesan Python):

In [17]:
import keyword
print(keyword.kwlist)

['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


### Jebakan: boleh, tapi bikin celaka
Nama function bawaan **bukan** keyword, jadi boleh dipakai sebagai nama variabel... tapi mesin aslinya ketimpa.

In [18]:
max = 4250000          # Raka nyimpen omzet tertinggi
print(max)

4250000


In [19]:
print(max(3, 7, 5))

TypeError: 'int' object is not callable

Perbaiki dengan `del` (menghapus variabel buatan kita):

In [20]:
del max
print(max(3, 7, 5))

7


### Konvensi (PEP 8)
| Jenis | Gaya | Contoh |
|---|---|---|
| Variabel | `snake_case` | `omzet_jakarta` |
| Konstanta | `UPPER_CASE` | `TARGET_HARIAN` |
| Class (nanti) | `PascalCase` | `LaporanHarian` |

In [21]:
TARGET_HARIAN = 10_000_000   # konstanta: nilai yang nggak berubah
omzet_bandung = 2875000      # variabel: snake_case, nama bermakna
print(TARGET_HARIAN, omzet_bandung)

10000000 2875000


## Bab 5 · Ngitung Omzet
Jalankan **data kit** dulu. Angka ini dipakai sepanjang kelas.

In [22]:
# ===== DATA KIT KOPI SENJA (penjualan Senin, 5 Oktober 2026) =====
omzet_jakarta  = 4_250_000     # Rupiah
omzet_bandung  = 2_875_000
omzet_surabaya = 3_520_000

trx_jakarta  = 170             # jumlah transaksi
trx_bandung  = 125
trx_surabaya = 160

TARGET_HARIAN = 10_000_000     # target omzet gabungan 3 cabang
print("Data kit angka siap ☕")

Data kit angka siap ☕


### `int`, `float`, `complex`
Desimal pakai **titik**. Underscore di angka (`4_250_000`) cuma buat mata manusia.

In [23]:
trx_jakarta = 170              # int   : bilangan bulat
rata_rata = 23395.6            # float : bilangan desimal
print(type(trx_jakarta), type(rata_rata), type(omzet_jakarta))

z = 3 + 4j                     # complex: kenalan aja, jarang dipakai di analisis data
print(type(z))

<class 'int'> <class 'float'> <class 'int'>
<class 'complex'>


### Operator: kalkulator Raka
| Operator | Arti | Contoh |
|---|---|---|
| `+ - *` | tambah, kurang, kali | `25000 * 170` |
| `/` | bagi (**selalu float**) | `10 / 2` → `5.0` |
| `//` | bagi, dibulatkan ke bawah | `455 // 50` → `9` |
| `%` | sisa bagi | `455 % 50` → `5` |
| `**` | pangkat | `1.1 ** 3` |

In [24]:
total_omzet = omzet_jakarta + omzet_bandung + omzet_surabaya
print(total_omzet)

print(omzet_jakarta - omzet_bandung)   # selisih
print(25000 * 170)                     # harga x jumlah cup
print(10 / 2)                          # kok ada .0?

10645000
1375000
4250000
5.0


In [25]:
# Cup dikirim per kardus isi 50. Total 455 cup:
print(455 // 50)    # kardus penuh
print(455 % 50)     # sisa cup

9
5


In [26]:
# Kalau omzet Jakarta naik 10% tiap bulan, 3 bulan lagi berapa?
print(omzet_jakarta * 1.1 ** 3)          # ** dikerjakan duluan
print(round(omzet_jakarta * 1.1 ** 3))

5656750.000000002
5656750


### Built-in function untuk angka

In [27]:
print(abs(omzet_bandung - omzet_jakarta))                 # selisih tanpa minus
print(round(23395.6043))
print(round(23395.6043, 1))
print(max(omzet_jakarta, omzet_bandung, omzet_surabaya))
print(min(omzet_jakarta, omzet_bandung, omzet_surabaya))
print(int(3.99))                                          # dipotong, BUKAN dibulatkan!
print(float(170))

1375000
23396
23395.6
4250000
2875000
3
170.0


**Cara mancing sendiri:** nggak perlu hafal semua function. Di Colab/Jupyter ketik `round?`, atau pakai `help()` di mana saja:

In [28]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.

    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



### Boolean: lampu BUKA/TUTUP
Cuma dua nilai: `True` dan `False` (huruf depan **kapital**). `=` memasukkan ke toples, `==` **bertanya** "apakah sama?".

In [29]:
print(omzet_jakarta > omzet_surabaya)
print(total_omzet >= TARGET_HARIAN)    # target tercapai?
print(trx_bandung != 125)
print(type(True))

True
True
False
<class 'bool'>


In [30]:
print(true)

NameError: name 'true' is not defined

### `and`, `or`, `not`
Pertemuan berikutnya, boolean ini jadi **otak** program lewat `if`.

In [31]:
# Target tercapai DAN Bandung di atas 3 juta?
print(total_omzet >= TARGET_HARIAN and omzet_bandung > 3_000_000)

# Target tercapai ATAU Bandung di atas 3 juta?
print(total_omzet >= TARGET_HARIAN or omzet_bandung > 3_000_000)

print(not True)

False
True
False


## Bab 6 · Beresin Nama Menu
Tiga kasir menulis menu yang sama dengan cara berbeda. Buat komputer, itu **tiga menu berbeda**.

In [32]:
# ===== DATA KIT TEKS: nama menu terlaris versi tiap kasir (berantakan!) =====
menu_kasir_jkt = "  es kopi SUSU gula aren "
menu_kasir_bdg = "ES KOPI SUSU GULA AREN"
menu_kasir_sby = "Es Kopi susu Gula Aren   "

kode_trx = "TRX-20261005-JKT-0170"   # transaksi terakhir hari itu
print("Data kit teks siap ☕")

Data kit teks siap ☕


In [33]:
print(menu_kasir_jkt == menu_kasir_bdg)

False


### String: untaian karakter

In [34]:
nama_toko = "Kopi Senja"                 # kutip dua
slogan = 'Seduh rasa, pulang bahagia'    # kutip satu, sama saja
alamat = """Jl. Senja No. 7
Jakarta Selatan"""                        # kutip tiga: bisa banyak baris

print("Kopi" + " " + "Senja")            # gabung
print("=" * 30)                          # ulang 30 kali
print(len(nama_toko))                    # jumlah karakter (spasi dihitung)
print(alamat)

Kopi Senja
10
Jl. Senja No. 7
Jakarta Selatan


### Indexing: tiap karakter punya nomor
```
 karakter:  T   R   X   -   2   0   2   6   1   0   0   5   -   J   K   T   -   0   1   7   0
 index +:   0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20
 index -: -21 -20 -19 -18 -17 -16 -15 -14 -13 -12 -11 -10  -9  -8  -7  -6  -5  -4  -3  -2  -1
```

In [35]:
print(kode_trx[0])      # index mulai dari 0!
print(kode_trx[-1])     # minus = hitung dari belakang
print(len(kode_trx))

T
0
21


Panjangnya 21, jadi index terakhir adalah **20**:

In [36]:
print(kode_trx[21])

IndexError: string index out of range

### Slicing: `teks[start:stop]`
Ambil dari `start` sampai **sebelum** `stop`. Bayangkan pisau memotong **di antara** karakter.

In [37]:
print(kode_trx[0:3])     # TRX
print(kode_trx[:3])      # start kosong = dari awal
print(kode_trx[4:12])    # tanggal
print(kode_trx[4:8])     # tahun
print(kode_trx[13:16])   # kode cabang
print(kode_trx[-4:])     # stop kosong = sampai akhir

TRX
TRX
20261005
2026
JKT
0170


### Slicing dengan langkah: `teks[start:stop:step]`

In [38]:
angka = "0123456789"
print(angka[::2])      # loncat 2 (genap)
print(angka[1::2])     # mulai index 1, loncat 2 (ganjil)
print(kode_trx[::-1])  # dibalik!

02468
13579
0710-TKJ-50016202-XRT


### String itu *immutable*
Tidak bisa diubah sebagian. Kayak tulisan yang dicetak di gelas: mau ganti, cetak gelas baru.

In [39]:
kode_trx[0] = "X"

TypeError: 'str' object does not support item assignment

In [40]:
kode_baru = kode_trx.replace("JKT", "BDG")   # bikin string BARU
print(kode_baru)
print(kode_trx)                              # yang lama tetap

TRX-20261005-BDG-0170
TRX-20261005-JKT-0170


### String method: alat beres-beres teks
| Method | Fungsi |
|---|---|
| `.strip()` | buang spasi di kiri-kanan |
| `.lower()` / `.upper()` | huruf kecil / besar semua |
| `.title()` | Huruf Depan Kapital |
| `.replace(a, b)` | ganti a dengan b |
| `.split(pemisah)` | pecah jadi list |
| `.count(x)` | hitung kemunculan |
| `.startswith(x)` | diawali x? |
| `.center(n)` | taruh di tengah selebar n |

Cari method lain: ketik `menu.` lalu tekan **Tab** di Colab, atau `dir(str)`.

In [41]:
print("  latte  ".strip())
print("Latte".upper())
print("es kopi".title())
print(kode_trx.split("-"))
print("latte".count("t"))
print(kode_trx.startswith("TRX"))

latte
LATTE
Es Kopi
['TRX', '20261005', 'JKT', '0170']
2
True


**Momen "aha"**: method bisa **dirantai**, dan `False` tadi berubah jadi `True`:

In [42]:
bersih_jkt = menu_kasir_jkt.strip().title()
bersih_bdg = menu_kasir_bdg.strip().title()
bersih_sby = menu_kasir_sby.strip().title()

print(bersih_jkt)
print(bersih_jkt == bersih_bdg and bersih_bdg == bersih_sby)

Es Kopi Susu Gula Aren
True


**Jebakan, tebak output:** (A) `[LATTE]` · (B) `[  LATTE  ]` · (C) `[  latte  ]`

In [43]:
menu = "  latte  "
menu.upper()
print("[" + menu + "]")

[  latte  ]


Method **tidak mengubah** aslinya, tapi mengembalikan string **baru**. Kalau mau disimpan:

In [44]:
menu = menu.upper()
print("[" + menu + "]")

[  LATTE  ]


### Casting: ganti jenis bahan
Data dari file atau input sering datang sebagai **teks**, walaupun isinya angka.

In [45]:
nomor = kode_trx[-4:]
print(nomor, type(nomor))

nomor_int = int(nomor)
print(nomor_int, type(nomor_int))   # nol di depan hilang

print(str(170) + " transaksi")
print(float("23.5"))

0170 <class 'str'>
170 <class 'int'>
170 transaksi
23.5


Kasir mengetik huruf **O**, bukan angka **0**:

In [46]:
print(int("17O"))

ValueError: invalid literal for int() with base 10: '17O'

## Bab 7 · Wadah Banyak Barang
### `list`: antrean pesanan
Urut, bisa diubah, bisa ditambah. Indexing & slicing **sama persis** seperti string.

In [47]:
cabang = ["Jakarta", "Bandung", "Surabaya"]
omzet  = [4_250_000, 2_875_000, 3_520_000]

print(omzet[0])
print(omzet[-1])
print(cabang[0:2])

print(len(omzet))
print(sum(omzet))
print(max(omzet), min(omzet))
print(sorted(omzet, reverse=True))   # urut dari terbesar

4250000
3520000
['Jakarta', 'Bandung']
3
10645000
4250000 2875000
[4250000, 3520000, 2875000]


In [48]:
omzet[1] = 2_900_000          # koreksi data Bandung: list BISA diubah
cabang.append("Yogyakarta")   # buka cabang baru
print(omzet)
print(cabang)

[4250000, 2900000, 3520000]
['Jakarta', 'Bandung', 'Surabaya', 'Yogyakarta']


### `tuple`: data yang dikunci

In [49]:
JAM_OPERASIONAL = (7, 22)        # buka jam 7, tutup jam 22
LOKASI_JKT = (-6.2, 106.8)       # koordinat
print(JAM_OPERASIONAL[0], LOKASI_JKT)

7 (-6.2, 106.8)


In [50]:
JAM_OPERASIONAL[0] = 8

TypeError: 'tuple' object does not support item assignment

### `set`: hanya yang unik (tanpa urutan)

In [51]:
menu_terjual = ["kopi susu", "latte", "kopi susu", "americano", "latte"]
menu_unik = set(menu_terjual)
print(menu_unik)          # urutan bisa beda-beda
print(len(menu_unik))

menu_tetap = frozenset(["kopi susu", "latte"])   # set yang dikunci (kenalan aja)
print(menu_tetap)

{'kopi susu', 'americano', 'latte'}
3
frozenset({'kopi susu', 'latte'})


### `dict`: papan menu (kunci → nilai)

In [52]:
harga = {
    "kopi susu": 25000,
    "latte": 28000,
    "americano": 22000,
}
print(harga["latte"])
harga["matcha latte"] = 30000   # tambah menu baru
print(harga)

28000
{'kopi susu': 25000, 'latte': 28000, 'americano': 22000, 'matcha latte': 30000}


In [53]:
print(harga["teh tarik"])

KeyError: 'teh tarik'

In [54]:
omzet_cabang = {"Jakarta": 4_250_000, "Bandung": 2_875_000, "Surabaya": 3_520_000}
print(omzet_cabang["Bandung"])
print(sum(omzet_cabang.values()))

2875000
10645000


| Wadah | Tanda | Urut? | Bisa diubah? | Boleh duplikat? |
|---|---|---|---|---|
| `list` | `[ ]` | ✅ | ✅ | ✅ |
| `tuple` | `( )` | ✅ | ❌ | ✅ |
| `set` | `{ }` | ❌ | ✅ | ❌ |
| `dict` | `{k: v}` | ✅ | ✅ | kunci ❌ |

## Bab 8 · Laporan buat Bu Sari
### `print()` lebih jauh & escape character
| Escape | Arti |
|---|---|
| `\n` | baris baru |
| `\t` | tab |
| `\"` `\'` | tanda kutip di dalam teks |
| `\\` | garis miring terbalik |

In [55]:
print("Jakarta", "Bandung", "Surabaya", sep=" | ")
print("Laporan \"Kopi Senja\"\nTanggal:\t05-10-2026")

Jakarta | Bandung | Surabaya
Laporan "Kopi Senja"
Tanggal:	05-10-2026


### Teks + angka = ?

In [56]:
total_omzet = 10645000
print("Total omzet: " + total_omzet)

TypeError: can only concatenate str (not "int") to str

### Evolusi string formatting (semua hasilnya sama)

In [57]:
print("Total omzet: " + str(total_omzet))        # 1. casting manual  → ribet
print("Total omzet: %d" % total_omzet)           # 2. gaya %  (jadul)
print("Total omzet: {}".format(total_omzet))     # 3. .format()
print(f"Total omzet: {total_omzet}")             # 4. f-string ⭐ pakai ini!

Total omzet: 10645000
Total omzet: 10645000
Total omzet: 10645000
Total omzet: 10645000


### Kekuatan super f-string
| Format | Contoh | Hasil |
|---|---|---|
| Ekspresi | `f"{total_omzet / 455}"` | hasil hitungan |
| Pemisah ribuan | `f"{total_omzet:,}"` | `10,645,000` |
| 2 desimal | `f"{106.4512:.2f}"` | `106.45` |

In [58]:
persen_target = total_omzet / TARGET_HARIAN * 100

print(f"Rata-rata   : {total_omzet / 455}")
print(f"Total omzet : Rp{total_omzet:,}")                    # gaya Inggris
print(f"Total omzet : Rp{total_omzet:,}".replace(",", "."))  # gaya Indonesia!
print(f"Capaian     : {persen_target:.2f}%")
print(f"Tercapai?   : {total_omzet >= TARGET_HARIAN}")

Rata-rata   : 23395.604395604394
Total omzet : Rp10,645,000
Total omzet : Rp10.645.000
Capaian     : 106.45%
Tercapai?   : True


### `input()`: biar laporannya bisa diisi
Di Colab muncul kotak isian. Ketik jawaban, lalu tekan **Enter**.

In [59]:
nama_analis = input("Nama analis: ")
print(f"Laporan disusun oleh {nama_analis}")

Nama analis: Raka Pratama
Laporan disusun oleh Raka Pratama


**Tebak output** (diketik: `10`): (A) `20` · (B) `1010` · (C) Error

In [60]:
trx = input("Jumlah transaksi: ")
print(trx * 2)

Jumlah transaksi: 10
1010


`input()` **selalu** menghasilkan string. Ubah dulu pakai `int()`:

In [61]:
trx = int(input("Jumlah transaksi: "))
print(trx * 2)

Jumlah transaksi: 10
20


## Final · Laporan Harian v1.0
Semua potongan puzzle sudah lengkap. Kode lengkapnya ada di section **00 · Demo Final** dan di file [`kopi-senja/laporan.py`](../kopi-senja/laporan.py).

- ✍️ Kerjakan versi latihannya di `02_latihan_peserta.ipynb` (section *Final*)
- 🔑 Cek jawabanmu di `03_kunci_jawaban.ipynb`

### Bersambung... ☕
Raka masih punya PR:
1. Data omzet masih **diketik manual** → baca file Excel/CSV (pandas)
2. Kalau target **nggak tercapai**, mau kasih peringatan → `if` / `else`
3. Kalau cabangnya **30**, masa nulis 30 baris print? → loop (`for`)
4. Mau bikin **mesin sendiri** kayak `print()` → function buatan sendiri